src/box/lua/config/utils/schema.lua

```diff
@@ -460,21 +482,104 @@ local function validate_table_is_array(data, ctx)
 
     -- An empty array is a valid array.
     if key_count == 0 then
-        return
+        return true
     end
 
     -- Check that the array starts from 1 and has no holes.
     if min_key ~= 1 then
-        walkthrough_error(ctx, 'An array must start from index 1, ' ..
-            'got min index %d', min_key)
+        return false, 'An array must start from index 1, got min index %d',
+            min_key
     end
 
     -- Check that the array has no holes.
     if max_key ~= key_count then
-        walkthrough_error(ctx, 'An array must not have holes, got ' ..
-            'a table with %d integer fields with max index %d', key_count,
-            max_key)
+        return false, 'An array must not have holes, got a table with %d ' ..
+            'integer fields with max index %d', key_count, max_key
     end
+
+    return true
+end
+
+-- Whether the given record or map definitely can't accept a
+-- non-empty array.
+--
+-- Returns true if any non-empty array definitely can't pass the
+-- schema's validation.
+--
+-- Returns false if this property can't be guaranteed.
+local function never_accept_nonempty_array(schema)
+    assert(schema.type == 'record' or schema.type == 'map')
+
+    -- An array with a fixed max length may be represented by a
+    -- record with numeric fields.
+    --
+    -- A record should have at least a field with the numeric key
+    -- 1 to be able to accept a non-empty array.
+    --
+    -- Example:
+    --
+    --  | local s = schema.new('foo', schema.record({
+    --  |     [1] = schema.scalar({type = 'string'}),
+    --  |     [2] = schema.scalar({type = 'string'}),
+    --  |     [3] = schema.scalar({type = 'string'}),
+    --  | }))
+    --  |
+    --  | s:validate({'a', 'b', 'c'}) -- OK
+    if schema.type == 'record' then
+        return schema.fields[1] == nil
+    end
+
+    -- If map's keys accept numeric values, then it is possible
+    -- that the map can accept a non-empty array value.
+    --
+    -- Example:
+    --
+    --  | local s = schema.new('foo', schema,map({
```

Comment:
There is a typo here, it should be `schema.map`: a dot instead of a comma.

Reply:
Eagle eye! Fixed in the code comment, in the commit message and in the pull request description.